### PR TITLE
refactor(neon_framework): Remove url launcher hack

### PR DESCRIPTION
**Pull request by provokateurin**
Closes https://github.com/nextcloud/neon/issues/2349

Using url_launcher directly is not possible because the URI needs to be complete because we often only get the path without the host etc.

### CODE REVIEW

**Comment by provokateurin** [comment]
Always using a real Account ensures the `completeUri` functionality is properly used. Less mocks, less problems in the real world.

File: packages/neon_framework/lib/src/testing/mocks.dart [hunk 1]
```diff
@@ -23,10 +22,17 @@ import 'package:shared_preferences_platform_interface/shared_preferences_platfor
...
+  Uri? serverURL,
+  String? username,
+  String? password,
+}) {
+  return Account(
+    (b) => b
+      ..serverURL = serverURL ?? Uri.parse('https://cloud.example.com:8443/nextcloud')
+      ..username = username ?? 'username'
+      ..password = password ?? 'password',
+  );
```
Comment (Leptopoda) [inline packages/neon_framework/lib/src/testing/mocks.dart:35]: Please add the default username and password value to the parameter list.

I'm also not a huge fan of this common code. I can see that you don't want to duplicate it over and over again, but I prefer unit tests to be independent.

File: packages/neon_framework/lib/src/pages/login_flow.dart [hunk 2]
```diff
@@ -39,10 +38,7 @@ class _LoginFlowPageState extends State<LoginFlowPage> {
 
     initSubscription = bloc.init.listen((result) async {
       if (result.hasData) {
-        await launchUrlString(
-          result.requireData.login,
-          mode: LaunchMode.externalApplication,
```
Comment (Leptopoda) [inline packages/neon_framework/lib/src/pages/login_flow.dart:44]: Why did you change the launch mode?
My last information was that it is preferred to use the browser, as webview might have issues with proxy setups

File: packages/neon_framework/lib/src/utils/launch_url.dart [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import 'package:flutter/cupertino.dart';
+import 'package:neon_framework/models.dart';
+import 'package:neon_framework/utils.dart';
+import 'package:url_launcher/url_launcher.dart' as url_launcher;
+
+Future<bool> launchUrl(BuildContext context, String url) async {
+  final account = NeonProvider.of<Account>(context);
+  final uri = account.completeUri(Uri.parse(url));
+  return url_launcher.launchUrl(uri);
+}
```
Comment (Leptopoda) [inline packages/neon_framework/lib/src/utils/launch_url.dart:10]: I'm not a big fan of this.
Passing around the BuildContext into utility functions.

I'd rather have this as an extension to the account itself.

**Review by Leptopoda — CHANGES_REQUESTED** [review]

File: packages/neon_framework/lib/src/utils/launch_url.dart (hunk 3 above)
Comment (provokateurin) [reply packages/neon_framework/lib/src/utils/launch_url.dart:10]: `Account.launchUrl(...)` doesn't make sense to me. Why is the Account launching the URL? We only need the Account to ensure the URL is valid which is a different thing.

File: packages/neon_framework/lib/src/pages/login_flow.dart (hunk 2 above)
Comment (provokateurin) [reply packages/neon_framework/lib/src/pages/login_flow.dart:44]: I forgot to change it

File: packages/neon_framework/lib/src/testing/mocks.dart (hunk 1 above)
Comment (provokateurin) [reply packages/neon_framework/lib/src/testing/mocks.dart:35]: I think this is better because it ensures we always use a URL that is special.
This doesn't make tests depending on each other and I don't see a downside in this shared mock account.

File: packages/neon_framework/lib/src/utils/launch_url.dart (hunk 3 above)
Comment (Leptopoda) [reply packages/neon_framework/lib/src/utils/launch_url.dart]: This is a bad timing for this kind of discussion.
In general, I'd say this should be handled by a bloc or for now in the widget itself.

It isn't obvious to the caller that this method needs access to the `Account` and therefore requires it to be in injected into the tree.
Maybe just make the account a parameter instead of the context and let the resolving happen outside this method.

We can discuss letting this is for now knowing that blocs will come to properly fix this.

Comment (provokateurin) [reply packages/neon_framework/lib/src/utils/launch_url.dart]: Ok, I'll go with the account as a parameter.

File: packages/neon_framework/lib/src/pages/login_flow.dart (hunk 2 above)
Comment (Leptopoda) [reply packages/neon_framework/lib/src/pages/login_flow.dart:44]: Still the same.

Comment (provokateurin) [reply packages/neon_framework/lib/src/pages/login_flow.dart:44]: No, the method that wraps it always uses LaunchMode.externalApplication

**Review by Leptopoda — APPROVED** [review]
